### ISSUE TITLE
The number of lines changes when moving between monitors of different DPI

### ISSUE THREAD

**Issue by page-down** [kovidgoyal/kitty#4114]
**Describe the bug**
Under macOS, moving windows between displays of different DPI works perfectly, except that the number of lines changes (increase or decrease) , even if the size of the window does not change.

The following is the case of integer scaling (2x), and fractional scaling is not tested.

**To Reproduce**

With the default font size of 11.0, the blank area increases a few pixel gradually as the window size increases (by option `initial_window_height`, not resizing window) after moving to a HiDPI monitor. When the window height is 24c, there is exactly one extra line.

1. Run kitty on a 1080p display, using the default font size of 11.0 and a height of 24c.
```shell
kitty --config=NONE -o font_size=11.0  -o initial_window_height=24c -o initial_window_width=80c -o remember_window_size=no -o "shell zsh -il"
```
2. Drag the window to the HiDPI display and an extra blank line will appear.
3. Press Enter and this empty line can render the text, for a total of 25c height.
4. Drag the window back to the low DPI display, which returns to 24c height.

When using 13.0 font size, JetBrains Mono font, 40c height, lines are missing after dragging the window to the HiDPI monitor.

1. Run kitty on a 1080p display.
```shell
kitty --config=NONE -o "font_family JetBrains Mono" -o font_size=13.0 -o initial_window_height=40c -o initial_window_width=80c -o remember_window_size=no -o "shell zsh -il" 
```
2. Dragging the window to the HiDPI display, when calculating the number of available lines, kitty thinks there is not enough space to display a full line, so blank area appears. Two lines are missing at this point. Only 38c height remains.
3. After dragging the window back to the low DPI display, the 40c height is restored. However, there are two more empty lines and they are always present in the scrollback and can be scrolled.

**Screenshots**

default font size 11.0, 24c height

1. 1080p display
![kitty_height_24c_1080p](https://user-images.githubusercontent.com/57713011/136897447-fba700af-1813-475e-976f-30af5f704110.png)

2. move to 4K HiDPI display
<img width="181" alt="kitty_moved_to_4k_same_window_size_25c" src="https://user-images.githubusercontent.com/57713011/136897472-7e082697-255b-46c4-8176-48cdc085cd86.png">

3. press Enter
<img width="202" alt="kitty_4k_empty_line_25" src="https://user-images.githubusercontent.com/57713011/136897533-f83c4166-8109-4e3c-859e-0e41844d2e25.png">

4. move back to 1080p display
![kitty_move_back_to_1080p_restored_24c](https://user-images.githubusercontent.com/57713011/136897575-ba82e3f5-2631-4d11-ba9e-7a471850e7d4.png)

font JetBrains Mono, size 13.0, 40c height

1. 1080p display
![kitty_font_size_13_height_40c_1080p](https://user-images.githubusercontent.com/57713011/136897584-ddfba82c-7ff5-4ea8-818d-c51b1c9b9032.png)

2. move to 4K HiDPI display
<img width="187" alt="kitty_font_size_13_4k_only_38c" src="https://user-images.githubusercontent.com/57713011/136897600-6a0a0026-39da-406e-b3d8-7384feb5eeb2.png">

3. move back to 1080p display
![kitty_font_size_13_move_back_to_1080p_40c_extra_empty_lines](https://user-images.githubusercontent.com/57713011/136897607-3357f554-dcce-42b3-a7ef-9b6eff3b52aa.png)

**Environment details**
- macOS 11.6
- kitty from master branch
- display 1: 1920 * 1080
- display 2: 3840 * 2160 @ 2x

**Additional context**
The above is only tested under macOS, not sure if similar problem exists under Linux.

When dragging between different DPI monitors, the resulting number of lines should remain the same when calculating the available rendering space.

**Comment by kovidgoyal** [comment]
cell size depends on font size and DPI, there is no guarantee that cell
size will remains the same when dpi changes. Since both font size and
DPI are floating point, and cell size in integer because of rounding,
cell sizes may change. Not to mention that other metrics in the window
such as border thickness and margins/paddings can also be dpi dependent.

**Comment by page-down** [comment]
> https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.resize_in_steps

Thank you, this configuration option is very useful.

However, I found that when used with `window_resize_step_lines 1`, every time the window size is changed on HiDPI displays, it changes by two lines per step.

```shell
kitty --config=NONE -o remember_window_size=no -o resize_in_steps=yes -o window_resize_step_cells=1 -o window_resize_step_lines=1
```
![kitty_resize_in_steps](https://user-images.githubusercontent.com/57713011/136909057-98b54bcb-ca65-4c3e-a384-548bcfd138ed.gif)

This option works fine on low DPI displays, adjusting one line per step.

**Comment by kovidgoyal** [comment]
Also for the resizing by two lines problem check in os_window_update_size_increments() in glfw.c it might be that on macOS the value needs to be scaled by the window scale.

**Comment by page-down** [comment]
> Oh and if you want to resize OS windows accurately use the kitty remote control command for it ...

Yes, thanks for the reminder, I currently have a shortcut set to reset the OS window size.

It would be nice if kitten could be executed automatically via hook, so that it could be called in case of DPI changes, etc.

However, I don't think it makes much sense to do hook specifically for this purpose.

> Also for the resizing by two lines problem check in os_window_update_size_increments() in glfw.c it might be that on macOS the value needs to be scaled by the window scale.

Yes, after scaling the values, changing the window size will be resized according to the single cell size.

```patch
diff --git a/kitty/glfw.c b/kitty/glfw.c
index 3a12411e..fe391026 100644
--- a/kitty/glfw.c
+++ b/kitty/glfw.c
@@ -807,8 +807,15 @@ create_os_window(PyObject UNUSED *self, PyObject *args, PyObject *kw) {
 void
 os_window_update_size_increments(OSWindow *window) {
     if (OPT(resize_in_steps)) {
-        if (window->handle && window->fonts_data) glfwSetWindowSizeIncrements(
-                window->handle, window->fonts_data->cell_width, window->fonts_data->cell_height);
+        if (window->handle && window->fonts_data) {
+            int cw = window->fonts_data->cell_width;
+            int ch = window->fonts_data->cell_height;
+#ifdef __APPLE__
+            if (window->viewport_x_ratio > 0) cw = (int)ceil(cw / window->viewport_x_ratio);
+            if (window->viewport_y_ratio > 0) ch = (int)ceil(ch / window->viewport_y_ratio);
+#endif
+            glfwSetWindowSizeIncrements(window->handle, cw, ch);
+        }
     } else {
         if (window->handle) glfwSetWindowSizeIncrements(
                 window->handle, GLFW_DONT_CARE, GLFW_DONT_CARE);

```

I can't test on Linux at the moment, but if there is no problem under Linux, is there a difference in the font size calculation?

**Comment by kovidgoyal** [comment]
On Tue, Oct 12, 2021 at 01:44:41AM -0700, page-down wrote:
> > Oh and if you want to resize OS windows accurately use the kitty remote control command for it ...
> 
> Yes, thanks for the reminder, I currently have a shortcut set to reset the OS window size.
> 
> It would be nice if kitten could be executed automatically via hook, so that it could be called in case of DPI changes, etc.
> 
> However, I don't think it makes much sense to do hook specifically for this purpose.

There already is: https://sw.kovidgoyal.net/kitty/launch/#watching-launched-windows
and in current master you can specify a watcher script in kitty.conf to
watch all windows.

**Comment by page-down** [comment]
> I have added a macOS specific fix.

Confirm that `resize_in_steps` is normal in HiDPI displays.

I found another problem.

For `resize-os-window`, under HiDPI monitor, after resizing by unit `cells`, the height is missing a line. The number of lines is correct in low DPI displays.

```shell
kitty --config=NONE -o "map f1 remote_control resize-os-window --width=80 --height=24 --unit=cells"
```

There are only 23 lines after pressing `F1`. There is no blank margin in the os window, just missing one line.

> >  It would be nice if kitten could be executed automatically via hook, so that it could be called in case of DPI changes, etc.
> > However, I don't think it makes much sense to do hook specifically for this purpose.
>
> There already is ... and in current master you can specify a watcher script in kitty.conf to watch all windows.

Thank you. DPI changes can be watched after adding the `on_dpi_change` related code to the following file.
- `kitty/launch.py`
- `kitty/window.py`

```shell
kitty -o watcher=/path/to/watcher.py
```

```python
# watcher.py
def on_dpi_change(boss: BossType, window: 'Window', data: Dict[str, Any]) -> None:
    boss.resize_os_window(window.os_window_id, width=80, height=24, unit='cells')
```
However, this doesn't make much sense, I can't think of any usage scenarios for `on_dpi_change`.
Also, since the kitty window is watched instead of the os window, the above code may be executed multiple times.

**Comment by kovidgoyal** [comment]
On Tue, Oct 12, 2021 at 05:49:10AM -0700, page-down wrote:
> > I have added a macOS specific fix.
> 
> Confirm that `resize_in_steps` is normal in HiDPI displays.
> 
> I found another problem.
> 
> For `resize-os-window`, under HiDPI monitor, after resizing by unit `cells`, the height is missing a line. The number of lines is correct in low DPI displays.

That's probably because of titlebar height. Try hiding the titlebar and
see if it works.

> 
> ```shell
> kitty --config=NONE -o "map f1 remote_control resize-os-window --width=80 --height=24 --unit=cells"
> ```
> 
> There are only 23 lines after pressing `F1`. There is no blank margin in the os window, just missing one line.
> 
> > >  It would be nice if kitten could be executed automatically via hook, so that it could be called in case of DPI changes, etc.
> > > However, I don't think it makes much sense to do hook specifically for this purpose.
> >
> > There already is ... and in current master you can specify a watcher script in kitty.conf to watch all windows.
> 
> Thank you. DPI changes can be watched after adding the `on_dpi_change` related code to the following file.
> - `kitty/launch.py`
> - `kitty/window.py`
> 
> ```shell
> kitty -o watcher=/path/to/watcher.py
> ```
> 
> ```python
> # watcher.py
> def on_dpi_change(boss: BossType, window: 'Window', data: Dict[str, Any]) -> None:
>     boss.resize_os_window(window.os_window_id, width=80, height=24, unit='cells')
> ```
> However, this doesn't make much sense, I can't think of any usage scenarios for `on_dpi_change`.
> Also, since the kitty window is watched instead of the os window, the above code may be executed multiple times.

a DPI change will cause a resize too, and yes I dont think this is worth
doing.

**Comment by page-down** [comment]
> Fixed by: `8057c42`

Thanks, after updating to the latest commit, everything works fine with the `resize-os-window` remote control.

To conclude this issue
- The number of screen pixels is integer.
- After the font size converted to HiDPI corresponding pts, although without the fractional part, but the pixels are more fine. This means that in the case of 2x scale up, there may be a 1 pixel difference.
  - For example, the Menlo font at 11.0 pts requires only 25px cell height at 2x, while at 1x it requires 13px.
- Since the pixel scale is not exactly equal to the 2x integer scaling ( actually fractional scaling), it can happen that the window is either insufficient or excessive in height after 2x.

What I'm curious about is why on proprietary platforms like Windows and macOS, many native apps don't change their content position at all when dragged across different DPI displays. It looks like they didn't do such a fine control over the rendering of the content like kitty did, but just rendered at the highest DPI and then scaled down to the rest.
